### DISCUSSION TITLE
Localization of attribute names

### DISCUSSION THREAD

**Discussion by wout** [luckyframework/lucky#1773] [Feature requests & Ideas] [answered]
Not too long ago, we added the `I18nBackend`, which does a great job for translations. But one thing is missing. The attribute names themselves are not translated, which looks odd.

I think this can be implemented easily by changing the following method to look in the i18n backend first:

https://github.com/luckyframework/avram/blob/main/src/lucky/ext/avram/errors.cr#L17-L19

To:

```crystal
module Avram
  class InvalidOperationError < AvramError
    # ...

    def renderable_details : String
      "#{localized_invalid_attribute_name} #{validation_messages.first}"
    end

    private def localized_invalid_attribute_name
      Avram.settings.i18n_backend.get?("attribute_#{invalid_attribute_name}") ||
        invalid_attribute_name
    end

    # ....
  end
end
```

Note that `Avram::I18nBackend` will need to have a `get?` variant to make this work. Alternatively there could be a `get_with_default(key: Symbol, String, default: String) : String` instead of a `get?(key : String) : String?` method to avoid nilable values.  

There can be a naming convention for attributes translations, something like:

```crystal
{
  # ...
  validate_required:       "is vereist",
  validate_uniqueness_of:  "bestaat al",

  attribute_email:         "e-mail",
  attribute_password:      "wachtwoord",    
}
```

We could even take it further by allowing operation-specific attribute translations. The only change needed would be storing the `operation` argument as in the instance variable:

https://github.com/luckyframework/avram/blob/main/src/avram/errors.cr#L62

```crystal
module Avram
  class InvalidOperationError < AvramError
    getter errors : Hash(Symbol, Array(String))
    private getter operation : Avram::Operation

    def initialize(@operation)
      #...
    end
  end
end
```

Then we could:

```crystal
module Avram
  class InvalidOperationError < AvramError
    # ...

    def renderable_details : String
      "#{localized_invalid_attribute_name} #{validation_messages.first}"
    end

    private def localized_invalid_attribute_name
      Avram.settings.i18n_backend.get?("attribute_#{operation.class.name.underscore}_#{name}") ||
        Avram.settings.i18n_backend.get?("attribute_#{name}") ||
        invalid_attribute_name
    end

    # ...
  end
end
```

That would be the cherry on the cake. :slightly_smiling_face: 

What do you guys think?

**Comment by jwoertink** [comment, marked as answer]
oh interesting. I guess I didn't consider the actual attribute names, but that makes sense. Het is een goed idee!

**Reply by wout** [reply]
Geweldig! ;)

I already have it working with Rosetta in one of my projects. I'll prepare a PR.

**Comment by wout** [comment]
I'm testing this in a few projects to iron out any possible issues. Once I feel comfortable enough with it, I'll create a PR.

**Reply by zw963** [reply]
Does this available for avram for now? (current avram version 1.2.0)

I need column name to I18n, for now, only message displayed by Chinese, it is wired.

**Reply by wout** [reply]
Not yet. Unfortunately, I got stuck somewhere. I have to rethink the strategy. But I need it soon myself for a project, so I'll pick this up again.

**Reply by wout** [reply]
@zw963 I'm back at this again, but I no longer see the issue. Would you mind helping me out here? Where do you have a mixture of attribute names and errors in different languages? The cases I looked through now are the following.

**1. API errors**
In `renderable_details` for the API output from operation errors. This is something that can be fixed, but still needs some tweaking. 

**2. CLI errors**
This can be solved using the same solution as the API output. But this is an issue for developers only, never for end users, so how useful is it here?

**3. `Shared::FieldErrors` component**
Here we don't even need to change anything in Avram. To fix this, `Shared::FieldError` should be updated from this:

```crystal
class Shared::FieldErrors(T) < BaseComponent
  needs attribute : Avram::PermittedAttribute(T)

  # Customize the markup and styles to match your application
  def render
    unless attribute.valid?
      div class: "error" do
        text "#{label_text} #{attribute.errors.first}"
      end
    end
  end

  def label_text : String
    Wordsmith::Inflector.humanize(attribute.name.to_s)
  end
end
```

To this:

```crystal
class Shared::FieldErrors(T) < BaseComponent
  needs attribute : Avram::PermittedAttribute(T)
  needs label_text : String?

  # Customize the markup and styles to match your application
  def render
    unless attribute.valid?
      div class: "error" do
        text "#{label_text_with_fallback} #{attribute.errors.first}"
      end
    end
  end

  def label_text_with_fallback
    label_text || Wordsmith::Inflector.humanize(attribute.name.to_s)
  end
end
```

This is something that probably needs to be changed in Lucky anyway because it's a bit odd to have a custom label for `Shared::Field`, and not reuse the same value for `Shared::FieldError`. So there's potentially a mismatch between the label on the field and the humanized attribute name.

Thoughts?

**Reply by zw963** [reply]
Hi, following is my attemption:

```crystal
# config/database.cr

struct MyI18n < Avram::I18nBackend
  def get(key : String | Symbol) : String
    {
      validate_confirmation_of:    "必须匹配",
      validate_exact_size_of:      "长度必须为 %d 个字符",
      validate_numeric_nil:        "数字不可以为空",
      validate_required:           "需要的",
      validate_uniqueness_of:      "已经存在了",
    }[key]
  end
end

Avram.configure do |settings|
  # Always parse `Time` values with these specific formats.
  # Used for both database values, and datetime input fields.
  # settings.time_formats << "%F"
  settings.i18n_backend = MyI18n.new
end
```

Then, when one column named "foo" is required, will print: "foo 需要的“，it's odd.

**Reply by wout** [reply]
Thanks! Could you elaborate a bit more one where and how that's rendered? Is that through the `shared/field_errors.cr` component generated by Lucky?

**Reply by zw963** [reply]
> Thanks! Could you elaborate a bit more one where and how that's rendered? Is that through the `shared/field_errors.cr` component generated by Lucky?

I think so, it happen in a full lucky project user login error message.

**Reply by wout** [reply]
Okay, then I know enough.

**Reply by wout** [reply]
It is now possible to translate attributes with Rosetta 0.13.0. Here's how to implement it.

Add the `avram.attribute_variants` key to your avram locale file and add all the attributes you want to see translated:
```yaml
en:
  avram:
    attribute_variants:
      email: email
   # ...
```

Update the `Shared::FieldError#label_text` method:
```crystal
# src/components/shared/field_error.cr
  # ...
  def label_text
    Rosetta.find(
      "avram.attribute_variants",
      Wordsmith::Inflector.humanize(attribute.name.to_s)
    ).t(variant: attribute.name.to_s)
  end
  # ...
```

This will try to look up the attribute translation and if it does not exist, it will use the attribute name as a fallback. You could remove the fallback, but then you'll get a runtime error if the translation could not be found.

**Reply by zw963** [reply]
Hi, do a quick test, but get a error like following:

web          |  7 | Rosetta::Backend.load("./config/rosetta")
web          |                       ^---
web          | Error: No pluralization rule is defined for "zh_CN"

https://github.com/crystal-china/website/commit/5f4e2c345f99284a5baceb866a588db90548a7f8

Any idea?

**Reply by wout** [reply]
Ah, that's probably a bug. The underscore in the symbol doesn't get converted to a hyphen so the pluralization rule is not found:

https://github.com/wout/rosetta/blob/main/src/rosetta/pluralization.cr#L155

If you use the hyphenated string value `"zh-CN"` instead, it should work. But I'll fix that straight away. Thanks for reporting!

**Reply by wout** [reply]
Done, this should be fixed in v0.13.2.

**Reply by zw963** [reply]
> Done, this should be fixed in v0.13.2.

Use `zh_CN` still not for me : (

https://github.com/crystal-china/website/commit/8e2e1711887c7b6cb6f66d40bc535833331c0c39


> If you use the hyphenated string value "zh-CN" instead, it should work.

I get it from the ` cat /etc/locale.gen` in a linux system, it's format like this: `zh_CN.UTF-8`

**Reply by wout** [reply]
What's happening exactly? Do you get an error or does it always fall back to `en`? If it's the latter, I'll have to make another patch to make Rosetta more tolerant to the provided locale values so it converts `zh_CN.UTF-8` to `zh-CN`.

**Reply by zw963** [reply]
> What's happening exactly? Do you get an error or does it always fall back to `en`? If it's the latter, I'll have to make another patch to make Rosetta more tolerant to the provided locale values so it converts `zh_CN.UTF-8` to `zh-CN`.

Same as before when use `zh_CN` as key.

```
web          |  7 | Rosetta::Backend.load("./config/rosetta")
web          |                       ^---
web          | Error: No pluralization rule is defined for "zh_CN"
```

But it works after rename to to `zh-CN`, localization of attribute names is work well too!

**Reply by zw963** [reply]
One more question, I saw the example in lucky doc, where to define the key `default.error.invalid_email`? 

What I mean is, I can't found the layer like `default -> error -> ???` in pre-defined yaml files, I can add this into whatever yaml files, right?

**Reply by wout** [reply]
This patch is now available in v0.13.3. You can give it a value with encoding and it will try to parse out a useful value:
```cr
Rosetta.locale = "zh_CN.UTF-8"
Rosetta.locale
# => "zh-CN"
```

**Reply by wout** [reply]
> Same as before when use `zh_CN` as key.

That's really weird. I've tried installing your app but I keep hitting a wall here:
```bash
lucky-1     | In lib/baked_file_system/src/baked_file_system.cr:183:8
lucky-1     | 
lucky-1     |  183 | {{ run("./loader", path, dir, include_dotfiles) }}
lucky-1     |           ^--
lucky-1     | Error: Error executing run (exit code: 1): ./loader /app/dist /app/config false
lucky-1     | 
lucky-1     | 
lucky-1     | stderr:
lucky-1     | 
lucky-1     |     Unhandled exception: path does not exist: /app/dist (BakedFileSystem::Loader::Error)
lucky-1     |       from /root/.cache/crystal/app-lib-baked_file_system-src-loader.cr/macro_run in '??'
lucky-1     |       from /root/.cache/crystal/app-lib-baked_file_system-src-loader.cr/macro_run in '__crystal_main'
lucky-1     |       from /root/.cache/crystal/app-lib-baked_file_system-src-loader.cr/macro_run in 'main'
lucky-1     |       from /lib/x86_64-linux-gnu/libc.so.6 in '??'
lucky-1     |       from /lib/x86_64-linux-gnu/libc.so.6 in '__libc_start_main'
lucky-1     |       from /root/.cache/crystal/app-lib-baked_file_system-src-loader.cr/macro_run in '_start'
lucky-1     |       from ???
lucky-1     | 
lucky-1 exited with code 1
```

> One more question, I saw the example in lucky doc, where to define the key `default.error.invalid_email`?
> 
> What I mean is, I can't found the layer like `default -> error -> ???` in pre-defined yaml files, I can add this into whatever yaml files, right?

You'll have to add this namespace in an existing file or create a separate file (e.g. `default.zh-CN.yml`). I tend to create a separate files for logical sections:

<img width="787" height="395" alt="image" src="https://github.com/user-attachments/assets/8a2a290b-8497-4033-9cf3-15588c68af98" />

But in the end it doesn't matter how you organise the locale files, whatever suites you best. The parser will merge all those files into one single hash and build the structure from there. (https://wout.github.io/rosetta/v0.13.0/locale_files/)

**Comment by zw963** [comment]
Oops, sorry, I don't know why, I missed this replies here.

> Error: Error executing run (exit code: 1): ./loader /app/dist /app/config false

Probably need create a empty `dist` folder before the first run of this app?  you can try `yarn watch`, wait it finished, then terminate it, run `lucky dev` again.

But, anyway, I add rosetta into my website, it work well if use `:"zh-CN"` as key, altough, zh_CN still not work, maybe you can remove the workaround for use with `zh_CN` to make code more simple, just need describe it in README is enough.

**Reply by wout** [reply]
Good idea, I'll update the documentation.
